File: codes/mobile/ios/Timia/Features/Schedule/ScheduleMapClusters.swift
import Foundation

let earthRadiusM = 6_371_000.0
let scheduleMapClusterRadiusM = 30.0

struct ScheduleMapTaskCluster: Identifiable, Equatable {
    let id: String
    let items: [ScheduleMapItem]
    let locationLat: Double
    let locationLng: Double
    let placeTitle: String
}

func haversineMeters(lat1: Double, lng1: Double, lat2: Double, lng2: Double) -> Double {
    let dLat = (lat2 - lat1) * .pi / 180
    let dLng = (lng2 - lng1) * .pi / 180
    let a = sin(dLat / 2) * sin(dLat / 2)
        + cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) * sin(dLng / 2) * sin(dLng / 2)
    return 2 * earthRadiusM * asin(min(1, sqrt(a)))
}

func sortScheduleMapClusterItems(_ items: [ScheduleMapItem]) -> [ScheduleMapItem] {
    let timed = items.filter(scheduleMapClusterIsTimed).sorted { left, right in
        let leftStart = left.startAt ?? ""
        let rightStart = right.startAt ?? ""
        if leftStart != rightStart { return leftStart < rightStart }
        let titleOrder = left.title.compare(right.title, locale: Locale(identifier: "zh"))
        if titleOrder != .orderedSame { return titleOrder == .orderedAscending }
        return left.id < right.id
    }
    return timed + items.filter { !scheduleMapClusterIsTimed($0) }
}

func scheduleMapClusterFocusIndex(_ items: [ScheduleMapItem], now: Date) -> Int {
    items.firstIndex { item in
        guard let date = scheduleMapParseISO(item.startAt) else { return false }
        return date >= now
    } ?? 0
}

func clusterScheduleMapItems(
    _ items: [ScheduleMapItem],
    now _: Date,
    placeFallback: String
) -> [ScheduleMapTaskCluster] {
    let buckets = groupScheduleMapItemsByCoordinate(items)
    var clusters: [(origin: (lat: Double, lng: Double), items: [ScheduleMapItem])] = []
    for bucket in buckets {
        guard let first = bucket.first else { continue }
        let origin = (lat: first.locationLat, lng: first.locationLng)
        var target: Int?
        var best = Double.greatestFiniteMagnitude
        for (index, cluster) in clusters.enumerated() {
            let fits = cluster.items.allSatisfy {
                haversineMeters(lat1: origin.lat, lng1: origin.lng, lat2: $0.locationLat, lng2: $0.locationLng)
                    <= scheduleMapClusterRadiusM
            }
            guard fits else { continue }
            let distance = haversineMeters(lat1: origin.lat, lng1: origin.lng, lat2: cluster.origin.lat, lng2: cluster.origin.lng)
            if distance < best {
                best = distance
                target = index
            }
        }
        if let target {
            clusters[target].items.append(contentsOf: bucket)
        } else {
            clusters.append((origin, bucket))
        }
    }
    return clusters.map { cluster in
        let sorted = sortScheduleMapClusterItems(cluster.items)
        let coord = majority(cluster.items) { scheduleMapCoordinateKey(lat: $0.locationLat, lng: $0.locationLng) }
        let names = cluster.items.compactMap { item -> String? in
            let trimmed = item.location?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return trimmed.isEmpty ? nil : trimmed
        }
        let placeTitle = names.isEmpty ? placeFallback : majority(names) { $0 }
        return ScheduleMapTaskCluster(
            id: scheduleMapCoordinateKey(lat: cluster.origin.lat, lng: cluster.origin.lng),
            items: sorted,
            locationLat: coord.locationLat,
            locationLng: coord.locationLng,
            placeTitle: placeTitle
        )
    }
}

private func scheduleMapClusterIsTimed(_ item: ScheduleMapItem) -> Bool {
    guard let startAt = item.startAt else { return false }
    return !startAt.isEmpty
}

private func majority<T>(_ values: [T], key: (T) -> String) -> T {
    var first: [String: T] = [:]
    var counts: [String: Int] = [:]
    var order: [String] = []
    for value in values {
        let group = key(value)
        if first[group] == nil {
            first[group] = value
            order.append(group)
        }
        counts[group, default: 0] += 1
    }
    var bestKey = order[0]
    var bestCount = -1
    for group in order {
        let count = counts[group] ?? 0
        if count > bestCount {
            bestKey = group
            bestCount = count
        }
    }
    return first[bestKey]!
}

File: codes/web/src/lib/scheduleMapClusters.ts
import { groupScheduleMapItemsByCoordinate, scheduleMapCoordinateKey } from "./scheduleMapPins.ts";

export const EARTH_RADIUS_M = 6_371_000;
export const SCHEDULE_MAP_CLUSTER_RADIUS_M = 30;

export type ScheduleMapCluster<T> = {
  id: string;
  items: T[];
  location_lat: number;
  location_lng: number;
  placeTitle: string;
};

type Clusterable = {
  id: string;
  title: string;
  location?: string | null;
  location_lat: number;
  location_lng: number;
  start_at?: string | null;
};

export function haversineMeters(
  a: { lat: number; lng: number },
  b: { lat: number; lng: number },
): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const sinLat = Math.sin(dLat / 2);
  const sinLng = Math.sin(dLng / 2);
  const h =
    sinLat * sinLat +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * sinLng * sinLng;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(h)));
}

export function sortScheduleMapClusterItems<T extends { id: string; title: string; start_at?: string | null }>(
  items: T[],
): T[] {
  const timed = items
    .filter((item) => item.start_at)
    .sort((left, right) => {
      const start = (left.start_at ?? "").localeCompare(right.start_at ?? "");
      if (start !== 0) return start;
      const title = left.title.localeCompare(right.title, "zh");
      if (title !== 0) return title;
      return left.id.localeCompare(right.id);
    });
  const undated = items.filter((item) => !item.start_at);
  return [...timed, ...undated];
}

export function scheduleMapClusterFocusIndex<T extends { start_at?: string | null }>(
  items: T[],
  now: Date,
): number {
  const nowMs = now.getTime();
  const upcoming = items.findIndex((item) => {
    if (!item.start_at) return false;
    return Date.parse(item.start_at) >= nowMs;
  });
  return upcoming >= 0 ? upcoming : 0;
}

function majority<T>(values: T[], keyOf: (value: T) => string): T {
  const first = new Map<string, T>();
  const counts = new Map<string, number>();
  for (const value of values) {
    const key = keyOf(value);
    if (!first.has(key)) first.set(key, value);
    counts.set(key, (counts.get(key) ?? 0) + 1);
  }
  let bestKey = [...first.keys()][0];
  let bestCount = -1;
  for (const [key, count] of counts) {
    if (count > bestCount) {
      bestKey = key;
      bestCount = count;
    }
  }
  return first.get(bestKey) as T;
}

export function clusterScheduleMapItems<T extends Clusterable>(
  items: T[],
  options: { now: Date; placeFallback: string },
): ScheduleMapCluster<T>[] {
  const buckets = groupScheduleMapItemsByCoordinate(items);
  const clusters: Array<{ origin: { lat: number; lng: number }; items: T[] }> = [];

  for (const bucket of buckets) {
    const origin = { lat: bucket[0].location_lat, lng: bucket[0].location_lng };
    let target = -1;
    let targetDistance = Number.POSITIVE_INFINITY;
    for (let index = 0; index < clusters.length; index += 1) {
      const fits = clusters[index].items.every(
        (member) =>
          haversineMeters(origin, { lat: member.location_lat, lng: member.location_lng }) <=
          SCHEDULE_MAP_CLUSTER_RADIUS_M,
      );
      if (!fits) continue;
      const distance = haversineMeters(origin, clusters[index].origin);
      if (distance < targetDistance) {
        target = index;
        targetDistance = distance;
      }
    }
    if (target >= 0) clusters[target].items.push(...bucket);
    else clusters.push({ origin, items: [...bucket] });
  }

  return clusters.map((cluster) => {
    const itemsSorted = sortScheduleMapClusterItems(cluster.items);
    const anchor = majority(cluster.items, (item) =>
      scheduleMapCoordinateKey(item.location_lat, item.location_lng),
    );
    const named = cluster.items
      .map((item) => (item.location ?? "").trim())
      .filter((name) => name.length > 0)
      .map((name) => ({ name }));
    const placeTitle = named.length
      ? majority(named, (row) => row.name).name
      : options.placeFallback;
    return {
      id: scheduleMapCoordinateKey(cluster.origin.lat, cluster.origin.lng),
      items: itemsSorted,
      location_lat: anchor.location_lat,
      location_lng: anchor.location_lng,
      placeTitle,
    };
  });
}
